### PR TITLE
feat: added memory hashing w/ padding handling

### PR DESCRIPTION
**Pull request by hackaugusto**
## Describe your changes

This add a procedure to hash memory between two pointers. This is the same that is needed to hash the peaks of an MMR into a single hash value.

## Checklist before requesting a review
- Repo forked and branch created from `next` according to naming convention.
- Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.

### CODE REVIEW

**Comment by hackaugusto** [comment]
Initially I had done this with a start_ptr + counter, but `mem_stream` increments the target address, so it felt less complicated to compare that against the target end_ptr.

File: miden/tests/integration/stdlib/crypto/rpo.rs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+#[test]
+fn test_hash_empty() {
+    // when the address range contains zero elements, the result is all zeros
+    let empty_range = "
+    use.std::crypto::hashes::rpo
+
+    begin
+        push.1000 # end address
+        push.1000 # start address
```
Comment (bobbinth) [inline miden/tests/integration/stdlib/crypto/rpo.rs:13]: Should this fail?

**Review by bobbinth — DISMISSED** [review]
Looks great! Thank you! I left a few small comments inline.

One thing I'm thinking about is the placement of this procedure. This procedure relies on the native hash function of the VM. If we change this, `rpo` namespace would not be correct any more. The other option is something like `std::crypto::hashes::native` - but I don't love this either.

Yet another option is to change this into an assembly instruction - something like `hdigest` (to be consistent with `hash`, `hperm`, `hmerge`). My hesitation here is that assembly instructions are usually bounded to some relatively small number of instructions (I think the biggest one can be close to 80 cycles or so) - but this one could result in a very long instruction sequence. But maybe that's OK?

File: miden/Cargo.toml [hunk 2]
```diff
@@ -56,6 +56,7 @@ serde_json = {version = "1.0.59", optional = true }
 stdlib = { package = "miden-stdlib", path = "../stdlib", version = "0.4", default-features = false }
 structopt = { version = "0.3", default-features = false, optional = true }
 verifier = { package = "miden-verifier", path = "../verifier", version = "0.5", default-features = false }
+vm-core = { package = "miden-core", path = "../core", version = "0.5", default-features = false }
```
[the remark this answers is not in the document]
Comment (hackaugusto) [reply miden/Cargo.toml:59]: that was a left over from some other changes I had done. moved it back

File: stdlib/asm/crypto/hashes/rpo.masm [hunk 3]
```diff
@@ -0,0 +1,66 @@
+#! Hashes the memory `start_addr` to `end_addr`.
+#!
+#! This requires that `end_addr=start_addr + 2n + 1`, otherwise the procedure will enter an infinite
+#! loop. `end_addr` is not inclusive.
+#!
+#! Stack transition:
+#! [A, B, C, start_addr, end_addr, ...] -> [A, B, C, end_addr, end_addr ...]
+#!
+#! Where `C` is the capacity word that will be used by the hashing function.
+export.hash_even_memory_addresses
```
[the remark this answers is not in the document]
Comment (hackaugusto) [reply stdlib/asm/crypto/hashes/rpo.masm:10]: done

File: stdlib/asm/crypto/hashes/rpo.masm [hunk 4]
```diff
@@ -0,0 +1,66 @@
+#! Hashes the memory `start_addr` to `end_addr`.
+#!
+#! This requires that `end_addr=start_addr + 2n + 1`, otherwise the procedure will enter an infinite
+#! loop. `end_addr` is not inclusive.
+#!
+#! Stack transition:
+#! [A, B, C, start_addr, end_addr, ...] -> [A, B, C, end_addr, end_addr ...]
```
[the remark this answers is not in the document]
Comment (hackaugusto) [reply stdlib/asm/crypto/hashes/rpo.masm:7]: done

File: stdlib/asm/crypto/hashes/rpo.masm [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+end
+
+#! Hashes the memory `start_addr` to `end_addr`, handles odd number of elements.
+#!
+#! This requires `end_addr > start_addr`, otherwise the procedure will enter an infinite loop.
+#! `end_addr` is not inclusive.
+#!
+#! Stack transition:
+#! [start_addr, end_addr, ...] -> [H, ...]
+export.hash_memory
```
[the remark this answers is not in the document]
Comment (hackaugusto) [reply stdlib/asm/crypto/hashes/rpo.masm:24]: added

File: stdlib/asm/crypto/hashes/rpo.masm [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  dup.13 dup.13 neq while.true
+    mem_stream
+    dup.13 dup.13 neq
+  end
+end
+
+#! Hashes the memory `start_addr` to `end_addr`, handles odd number of elements.
+#!
+#! This requires `end_addr > start_addr`, otherwise the procedure will enter an infinite loop.
+#! `end_addr` is not inclusive.
```
[the remark this answers is not in the document]
Comment (hackaugusto) [reply stdlib/asm/crypto/hashes/rpo.masm:20]: that would require the addresses to be u32, should I add it?

[the remark this answers is not in the document]
Comment (bobbinth) [reply stdlib/asm/crypto/hashes/rpo.masm]: Yes - in the future we'll try to add hard restrictions in the VM that addresses cannot be greater than u32 value.

File: miden/tests/integration/stdlib/crypto/rpo.rs (hunk 1 above)
Comment (bobbinth) [reply miden/tests/integration/stdlib/crypto/rpo.rs:13]: I think this is fine. I think I saw a comment somewhere that said that start address must be smaller than end address - if that's the case, we should modify that comment to say if the address are the same this just computes a hash of a single word.

File: stdlib/asm/crypto/hashes/rpo.masm (hunk 6 above)
[the remark this answers is not in the document]
Comment (hackaugusto) [reply stdlib/asm/crypto/hashes/rpo.masm]: done

File: miden/tests/integration/stdlib/crypto/rpo.rs (hunk 1 above)
Comment (hackaugusto) [reply miden/tests/integration/stdlib/crypto/rpo.rs:13]: done together w/ https://github.com/0xPolygonMiden/miden-vm/pull/805#discussion_r1152648652

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you! I left two small comments inline - after these are addressed, we can merge.

**Comment by bobbinth** [comment]
Oh - one other thing: could you rebase from `next` to make sure the tests are passing (the may break because of the test framework refactoring).

File: miden/tests/integration/stdlib/crypto/native.rs [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    // checks the hash of 1 is the same when using hash_memory
+    let one_element = "
+    use.std::crypto::hashes::native
+
+    begin
+        # insert 1 to memory
+        push.1.1000 mem_store
+
+        push.1002 # end address
+        push.1000 # start address
```
[the remark this answers is not in the document]
Comment (hackaugusto) [reply miden/tests/integration/stdlib/crypto/native.rs:105]: I added another test case to cover that scenario. this test was covering a single iteration without padding. The range `[1000,1001)` needs padding.

**Pull request merged**